src/pokemon.c

```diff
@@ -6941,3 +6942,198 @@ void UpdateDaysPassedSinceFormChange(u16 days)
+        {
+            type = GetMonData(mon, MON_DATA_TERA_TYPE);
+        }        
+        else if (gBattleMons[battler].type1 != TYPE_MYSTERY)
+        {
+            type = gBattleMons[battler].type1;
+        }
+        else if (gBattleMons[battler].type2 != TYPE_MYSTERY)
+        {
+            type = gBattleMons[battler].type2;
+        }
+        else if (gBattleMons[battler].type3 != TYPE_MYSTERY)
+        {
+            type = gBattleMons[battler].type3;
+        }
+    }
+
+    else if (gMovesInfo[move].effect == EFFECT_TERRAIN_PULSE 
+        && ((IsMonGrounded(heldItemEffect, ability, type1, type2) && gBattleMons[battler].species != species) 
+        || (IsBattlerTerrainAffected(battler, STATUS_FIELD_TERRAIN_ANY) && gBattleMons[battler].species == species)))
+    {  
+        if (gFieldStatuses & STATUS_FIELD_ELECTRIC_TERRAIN)
+        {
+            return TYPE_ELECTRIC;
+        }
+        else if (gFieldStatuses & STATUS_FIELD_GRASSY_TERRAIN)
+        {
+            return TYPE_GRASS;
+        }
+        else if (gFieldStatuses & STATUS_FIELD_MISTY_TERRAIN)
+        {
+            return TYPE_FAIRY;
+        }
+        else if (gFieldStatuses & STATUS_FIELD_PSYCHIC_TERRAIN)
+        {
+            return TYPE_PSYCHIC;
+        }
+        else //failsafe
+        {
+            type = TYPE_NORMAL;
+        }        
+    }
+
+    if (gMovesInfo[move].effect == EFFECT_WEATHER_BALL && WEATHER_HAS_EFFECT)
+    {
+        if (gBattleWeather & B_WEATHER_RAIN && heldItemEffect != HOLD_EFFECT_UTILITY_UMBRELLA)
+        {
+            return TYPE_WATER;
+        }
+        else if (gBattleWeather & B_WEATHER_SANDSTORM)
+        {
+            return TYPE_ROCK;
+        }
+        else if (gBattleWeather & B_WEATHER_SUN && heldItemEffect != HOLD_EFFECT_UTILITY_UMBRELLA)
+        {
+            return TYPE_FIRE;
+        }
+        else if (gBattleWeather & B_WEATHER_SNOW)
+        {
+            return TYPE_ICE;
+        }
+        else
+        {
+            return TYPE_NORMAL;
+        }  
+    }
+
+    if (ability == ABILITY_NORMALIZE 
+        && gMovesInfo[move].type != TYPE_NORMAL 
+        && GetActiveGimmick(battler) != GIMMICK_Z_MOVE)
+    {
+        type = TYPE_NORMAL;
+    }
+
+    if ((gFieldStatuses & STATUS_FIELD_ION_DELUGE 
+        && type == TYPE_NORMAL) 
+        || gStatuses4[battler] & STATUS4_ELECTRIFIED)
+    {
+        type = TYPE_ELECTRIC;
+    }
+
+    if (gMovesInfo[move].type == TYPE_NORMAL
+                && gMovesInfo[move].category != DAMAGE_CATEGORY_STATUS
+                && ((ability == ABILITY_PIXILATE       && (type = TYPE_FAIRY))
+                    || (ability == ABILITY_REFRIGERATE && (type = TYPE_ICE))
+                    || (ability == ABILITY_AERILATE    && (type = TYPE_FLYING))
+                    || ((ability == ABILITY_GALVANIZE) && (type = TYPE_ELECTRIC)))
+                    )
+    {
+        return type;
+    }
```

Comment:
for the sake of better readability I would add a depth here so it would look like this:
```
if (gMovesInfo[move].type == TYPE_NORMAL && gMovesInfo[move].category != DAMAGE_CATEGORY_STATUS)
{
    switch (ability)
    {
    case ABILITY_PIXILATE: return TYPE_FAIRY;
    case ABILITY_REFRIGERATE: return TYPE_ICE;
    case ABILITY_AERILATE: return TYPE_FLYING;
    case ABILITY_GALVANIZE: return TYPE_ELECTRIC;
    default: break;
    }
}
```